### PR TITLE
Allow forced save as JPEG

### LINKED ISSUE #15: provide a system for selecting filetype when an image is written
At the moment images that image_vise 'writes' are either PNGs or the filetype they began as when read into the system. It would be useful to add functionality for specifying a filetype (e.g. jpg) when the file is written.

### PR DESCRIPTION
**Pull request by bermannoah**
Closes #15.

This PR ballooned a bit... BUT it makes it possible to override the default PNG export with JPG and GIF options. It also adds a custom metadata field that we can eventually expand to add more modifiers. (For example, there's now an operator that can be used to set jpg quality.)

There's also an operator that adds a background_fill, especially useful for handling PNG to JPG conversions. There is support for doing this with one modifier in recent versions of rmagick/imagemagick but doing so makes older versions blow up, so: the two step process within the conversion.

### CODE REVIEW

File: spec/image_vise/image_preview_test_spec.rb [hunk 1]
```diff
@@ -29,22 +29,22 @@
       get image_request.to_path_params('1337ness')
       examine_image_from_string(last_response.body)
       expect(last_response.headers['Content-Type']).to eq('image/jpeg')
-      expect(last_response.headers['Content-Length']).to eq("70559")
+      expect(last_response.headers['Content-Length']).to be_between("60000","75000").inclusive
```
Comment (bermannoah) [inline spec/image_vise/image_preview_test_spec.rb]: It's not as specific as I would like but it _hopefully_ helps make it less necessary to have an environment exactly like mine to achieve passing tests. :)

File: lib/image_vise/render_engine.rb [hunk 2]
```diff
@@ -315,7 +318,8 @@ def apply_pipeline(source_file_path, pipeline, source_file_type, render_to_path)
     # Otherwise, keep the original format for as far as the supported formats list goes.
     render_file_type = PNG_FILE_TYPE if magick_image.alpha?
     render_file_type = PNG_FILE_TYPE unless output_file_type_permitted?(render_file_type)
-    
+    render_file_type = JPG_FILE_TYPE if magick_image["render_as"]
```
Comment (roysab) [inline lib/image_vise/render_engine.rb]: Maybe check for: `magick_image["render_as"] == 'jpg'`

Comment (bermannoah) [inline lib/image_vise/render_engine.rb]: 👍  added the more precise check - thanks!

File: lib/image_vise/render_engine.rb [hunk 3]
```diff
@@ -303,24 +309,39 @@ def enable_forking?
...
     render_file_type = source_file_type
-    
+
     # Load the first frame of the animated GIF _or_ the blended compatibility layer from Photoshop
     image_list = Magick::Image.read(source_file_path)
     magick_image = image_list.first
 
+    # Strip any incorrect or malicious data from our custom config field.
+    magick_image["image_vise_config_data"] = Hash.new.to_json
+
```
Comment (bermannoah) [inline lib/image_vise/render_engine.rb]: If there's a better time/place in the process to do this I'm happy to move it.

File: lib/image_vise/render_engine.rb [hunk 4]
```diff
@@ -303,24 +309,39 @@ def enable_forking?
...
+    custom_config_options = JSON.parse(magick_image["image_vise_config_data"])
     render_file_type = PNG_FILE_TYPE if magick_image.alpha?
     render_file_type = PNG_FILE_TYPE unless output_file_type_permitted?(render_file_type)
-    
+    render_file_type = JPG_FILE_TYPE if custom_config_options["filetype"] == 'jpg'
+    render_file_type = GIF_FILE_TYPE if custom_config_options["filetype"] == 'gif'
+    jpg_quality = custom_config_options["jpg_quality"] if custom_config_options["jpg_quality"]
+
+    # Remove our custom config info from image prior to write.
+    magick_image["image_vise_config_data"] = ""
```
Comment (bermannoah) [inline lib/image_vise/render_engine.rb]: There has got to be a better way to do this but rmagick doesn't support stripping individual properties, it's all or nothing.

File: lib/image_vise/operators/jpg_quality.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# Changes the export quality of jpg images.
+#
+# The corresponding Pipeline method is `jpg_quality`.
+class ImageVise::JpgQuality < Ks.strict(:jpg_quality)
+
+  def initialize(*)
+    super
+    self.jpg_quality = jpg_quality.to_s
```
Comment (julik) [inline lib/image_vise/operators/jpg_quality.rb]: Do you really need to convert it to a string here? ivise params are passed in a JSON blob effectively, and JSON supports just ints.

**Comment by julik** [comment]
@bermannoah Back to you ;-)

**Comment by bermannoah** [comment]
@julik 👍  👍  👍  I can't think of anything to change -- mind if I go ahead and merge?

**Pull request merged**